src/compiler/moduleSpecifiers.ts

```diff
@@ -369,13 +370,49 @@ namespace ts.moduleSpecifiers {
         return sortedPaths;
     }
 
-    function tryGetModuleNameFromAmbientModule(moduleSymbol: Symbol): string | undefined {
+    function tryGetModuleNameFromAmbientModule(moduleSymbol: Symbol, checker: TypeChecker): string | undefined {
         const decl = find(moduleSymbol.declarations,
             d => isNonGlobalAmbientModule(d) && (!isExternalModuleAugmentation(d) || !isExternalModuleNameRelative(getTextOfIdentifierOrLiteral(d.name)))
         ) as (ModuleDeclaration & { name: StringLiteral }) | undefined;
         if (decl) {
             return decl.name.text;
         }
+
+        // the module could be a namespace, which is default export from an ambient module.
+        /**
+         * declare module "m" {
+         *     namespace ns {
+         *         class c {}
+         *     }
+         *     export = ns;
+         * }
+         */
+        // `import {c} from "m";` is valid, in which case, `moduleSymbol` is "ns", but the module name should be "m"
+        const ambientModuleDeclareCandidates = map(moduleSymbol.declarations,

```

Comment:
```suggestion
        const ambientModuleDeclareCandidates = mapDefined(moduleSymbol.declarations,

```
should simplify the lines below a bit. (Namely you should just be able to pull the first element out, rather than `find`ing a defined element).